### PR TITLE
SONARJAVA-1300 resolve union types to least upper bound

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  private TypeResolutionHelper() {
+  }
+
+  /**
+   * Compute the "Least Upper Bound" ("lub", jls8 §4.10.4) of a list of type. is a shared supertype that is more specific than any
+   * other shared supertype (that is, no other shared supertype is a subtype of the least upper bound)
+   * @param types
+   * @return the least upper bound of the types
+   */
+  public static JavaType leastUpperBound(List<JavaType> types) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java]: This method should not rely on a List of JavaType but a list of Type.

File: java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    Set<JavaType> minimalErasedCandidates = minimalErasedCandidates(erasedCandidates);
+    if (minimalErasedCandidates.isEmpty()) {
+      return Symbols.unknownType;
+    }
+    Multimap<JavaType, ParametrizedTypeJavaType> relevants = relevant(minimalErasedCandidates, supertypes);
+    Map<JavaType, ParametrizedTypeJavaType> candidates = leastContainingParameterization(relevants);
+
+    List<JavaType> bests = best(minimalErasedCandidates, candidates);
+    if (bests.size() > 1) {
+      // FIXME should be the Best(W1) & ... & Best(Wr)
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    return new ParametrizedTypeJavaType(t1.rawType.symbol, substitution);
+  }
+
+  private static JavaType leastContainingTypeArgument(JavaType t1, JavaType t2) {
+    if (isInvalid(t1) || isInvalid(t2)) {
+      return Symbols.unknownType;
+    } else if (t1.equals(t2)) {
+      return t1;
+    }
+    // FIXME types of wildcard trees are not implemented
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java [hunk 4]
```diff
@@ -84,4 +90,107 @@ public boolean isSubtype(JavaType t, JavaType s) {
     return result;
   }
 
+  /**
+   * Compute the "Least Upper Bound" ("lub", jls8 §4.10.4) of a list of type. The "lub" is a shared supertype that is more specific than any
+   * other shared supertype (that is, no other shared supertype is a subtype of the least upper bound)
+   *
+   * FIXME Parameterized types are currently ignored, as the method is used only to handle Union Types Trees, themselves being used only
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java [hunk 5]
```diff
@@ -84,4 +90,107 @@ public boolean isSubtype(JavaType t, JavaType s) {
...
+      return Symbols.unknownType;
+    }
+
+    List<Type> candidates = intersection(supertypes);
+    List<Type> minimalCandidates = minimalCandidates(candidates);
+    if (minimalCandidates.isEmpty()) {
+      return Symbols.unknownType;
+    }
+
+    // FIXME huge approximation: should be the bound of all the minimalCandidates, not only the first type
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 6]
```diff
@@ -540,9 +543,12 @@ public void visitTypeCast(TypeCastTree tree) {
 
   @Override
   public void visitUnionType(UnionTypeTree tree) {
-    resolveAs((List<? extends Tree>) tree.typeAlternatives(), JavaSymbol.TYP);
-    //TODO compute type of union type: lub(alternatives) cf JLS8 14.20
-    registerType(tree, Symbols.unknownType);
+    resolveAs((List<TypeTree>) tree.typeAlternatives(), JavaSymbol.TYP);
+    ImmutableList.Builder<Type> builder = ImmutableList.builder();
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: builder is not the perfect name here. uniontype.

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java (hunk 4 above)
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: Do we want to track this with a FIXME ?

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 7]
```diff
@@ -695,6 +695,10 @@ private static boolean isProtectedAccessible(JavaSymbol symbol, JavaSymbol.TypeJ
     return true;
   }
 
+  Type leastUpperBound(List<Type> typeAlternatives) {
+    return types.leastUpperBound(typeAlternatives.toArray(new Type[typeAlternatives.size()]));
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: Why convulotions between list and array ?

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java [hunk 8]
```diff
@@ -84,4 +90,107 @@ public boolean isSubtype(JavaType t, JavaType s) {
...
+        return type;
+      } else if (result.isUnknown()) {
+        // save first interface
+        result = type;
+      }
+    }
+    return result;
+  }
+
+  private static List<List<Type>> supertypes(Type... types) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: Why a List of List ? why not a Set ?

File: java-squid/src/main/java/org/sonar/java/resolve/Types.java [hunk 9]
```diff
@@ -84,4 +90,107 @@ public boolean isSubtype(JavaType t, JavaType s) {
...
+   * @param types
+   * @return the least upper bound of the types
+   */
+  public Type leastUpperBound(Type... types) {
+    Preconditions.checkArgument(types.length > 1);
+
+    List<List<Type>> supertypes = supertypes(types);
+    if (containsParameterizedTypes(supertypes)) {
+      // Parameterized types are ignored, erased supertypes should be calculated
+      return Symbols.unknownType;
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Types.java]: This is actually deadcode regarding where it is used, so I would not return unknown but actually fail in this case if you want to cover it.

**Pull request merged**